services/perfLogger/perfLogger.ts

```diff
@@ -0,0 +1,155 @@
+
+        for await (const log of perfLogs) {
+            if(log.createdOn > currentEndTime) {
+                await this.generateMetrics(aggregatedLog, currentEndTime);
+                aggregatedLog = {};
+                currentEndTime = this.getStartAndEndEpochTime(log.createdOn).endTime;
+            }
+            if(!aggregatedLog[log.type]){
+                aggregatedLog[log.type] = [];
+            }
+            aggregatedLog[log.type].push(log);
+        }
+        this.generateMetrics(aggregatedLog, currentEndTime); // generate metrics for last day
+    }
+    private async generateMetrics(aggregatedLog, currentEndTime) {
+        if(_.isEmpty(aggregatedLog)) {
+            return;
+        }
+        // TODO: aggregate logs based on type and log metric event
+        await this.updateLastSyncDate(currentEndTime); // update last processed time
+    }
+    private async updateLastSyncDate(currentEndTime){
+        await this.settingSDK.put(LAST_PERF_LOG_PROCEEDED_ON, {lastProcessedOn: currentEndTime}); // should be un-commented
+    }
+    private getUnProcessedLogs({startTime, endTime}, limit){
+        const that = this;
+        const generatorFunction = async function* () {
+            let lastProcessedContent: any = {};
+            while(true) {
+                let logs = await that.getLogsFromDB({startTime, endTime}, {limit});
+                const lastProcessedContentIndex = _.findIndex(logs, { _id: lastProcessedContent._id});
+                if (lastProcessedContentIndex !== -1) {
+                    logs = logs.slice(lastProcessedContentIndex + 1); // slice off already processed data
+                }
+                if(!logs.length){
+                    return; // terminates generator 
+                }
+                for (const log of logs) {
+                    yield log;
+                }
+                lastProcessedContent = logs[logs.length - 1]; // used to remove already processed data
+                startTime = lastProcessedContent.createdOn; // update start time with last fetched rows createdOn
+            }
+        }
+        return {
+            [Symbol.asyncIterator]: generatorFunction
+        }
+    }
+    private async getLogsFromDB<T>({startTime, endTime}: {startTime: number, endTime: number},
+            {fields, limit}: { fields?: string[], limit?: number}): Promise<IPerfLog<T>[]> {
+        const query = {
+            selector: {
+                createdOn: {
+                    "$gte": startTime,
+                    "$lte": endTime
+                }
+            },
+            sort: ['createdOn']
+        }
+        if(limit){
+            query['limit'] = limit;
+        }
+        if (fields) {
+            query['fields'] = fields;
+        }
+        return this.dbSDK.find(DB_NAME, query).then(data => data.docs);
+    }
+    private getStartAndEndEpochTime(dateInEpoch = Date.now()): {startTime: number, endTime: number}{
```

Comment:
This is used for ?

Reply:
used for querying from database and aggregating log for each day